p4include/v1model.p4

```diff
@@ -88,27 +88,123 @@ enum MeterType {
+     * count, or both, depending upon the CounterType of the counter
+     * instance used when it was constructed.
+     *
+     * @param index The index of the counter state in the array to be
+     *              updated, normally a value in the range [0,
+     *              size-1].  If index >= size, no counter state will be
+     *              updated.
+     */
     void count(in bit<32> index);
 }
 
 extern direct_counter {
     direct_counter(CounterType type);
+    /***
+     * count() on a direct_counter object may only be called within an
+     * action executed as a result of matching a table entry, of a
+     * table that has a direct_counter associated with it.  Calling
+     * count() causes the counter state associated with the matched
+     * table entry to be read, modified, and written back, atomically
+     * relative to the processing of other packets.
+     *
+     * TBD: Is the direct_counter state updated even if an action is
+     * executed that does not call the count() method?
+     */
     void count();
 }
 
+#define V1MODEL_METER_COLOR_GREEN  0
+#define V1MODEL_METER_COLOR_YELLOW 1
+#define V1MODEL_METER_COLOR_RED    2
+
 extern meter {
     meter(bit<32> size, MeterType type);
+    /***
+     * execute_meter() causes the meter state with the specified index
+     * to be read, modified, and written back, atomically relative to
+     * the processing of other packets, and an integer encoding of one
+     * of the colors green, yellow, or red to be written to the result
+     * out parameter.
+     *
+     * @param index The index of the meter state in the array to be
+     *              updated, normally a value in the range [0,
+     *              size-1].  If index >= size, no meter state will be
+     *              updated.
+     * @param result Type T must be bit<W> with W >= 2.  When index is
+     *              in range, the value of result will be assigned 0
+     *              for color GREEN, 1 for color YELLOW, and 2 for
+     *              color RED (see RFC 2697 and RFC 2698 for the
+     *              meaning of these colors).  When index is out of
+     *              range, the final value of result is not specified,
+     *              and should be ignored by the caller.
+     */
     void execute_meter<T>(in bit<32> index, out T result);
 }
 
 extern direct_meter<T> {
     direct_meter(MeterType type);
+    /***
+     * read() may only be called within an action executed as a result
+     * of matching a table entry, of a table that has a direct_meter
+     * associated with it.  Calling read() causes the meter state
+     * associated with the matched table entry to be read, modified,
+     * and written back, atomically relative to the processing of
+     * other packets, and an integer encoding of one of the colors
+     * green, yellow, or red to be written to the result out
+     * parameter.
+     *
+     * TBD: Is the meter state updated even if an action is executed
```

Comment:
same answer as for direct counter above
in a way for direct meters, this is implied by the method name. The meter is always executed, but you only call read() if you want to output the value to a field, whereas for (indirect) meters, the method is called "execute_meter".

Reply:
Thanks.  I have another commit that will make similar changes to direct_meter comments as I mentioned above for direct_counter.